### PR TITLE
[RemoveDIs][DebugInfo][NFC] Add Instruction and convenience functions to DPValue

### PR DESCRIPTION
**Pull request by SLTozer**
This patch adds a set of functions to the DPValue class that conveniently perform some common operations, and some that replicate existing functions on `DbgVariableIntrinsic` and its subclasses. These functions aren't used in this patch, but will be used in patches to come (added in this patch to make review easier - they all perform trivial tasks or exactly replicate the behaviour of existing functions).

### CODE REVIEW

File: llvm/lib/IR/DebugProgramInstruction.cpp [hunk 1]
```diff
@@ -41,6 +41,36 @@ DPValue::DPValue(Metadata *Location, DILocalVariable *DV, DIExpression *Expr,
...
 void DPValue::deleteInstr() { delete this; }
 
+DPValue *DPValue::createDPValue(Value *Location, DILocalVariable *DV,
+                                DIExpression *Expr, const DILocation *DI,
+                                Instruction *InsertBefore) {
+  auto *NewDPValue = new DPValue(ValueAsMetadata::get(Location), DV, Expr, DI,
+                                 LocationType::Value);
+  if (InsertBefore)
+    InsertBefore->getParent()->insertDPValueBefore(NewDPValue,
+                                                   InsertBefore->getIterator());
```
Comment (jmorse) [inline llvm/lib/IR/DebugProgramInstruction.cpp:51]: Do you think it's worth having an iterator-taking version too, to match the equivalent insertBefore with BasicBlock::iterator? We don't do before-or-after debuginfo shenanigans here, but that might be necessary if you're looking to have source-equivalence between dbg.value and DPValue mode implementations of code.

File: llvm/lib/IR/DebugProgramInstruction.cpp [hunk 2]
```diff
@@ -328,6 +392,18 @@ void DPMarker::insertDPValue(DPValue *New, bool InsertAtHead) {
   StoredDPValues.insert(It, *New);
   New->setMarker(this);
 }
```
Comment (jmorse) [inline llvm/lib/IR/DebugProgramInstruction.cpp:394]: Nit, whitespace between functions desired.

**Review by jmorse — APPROVED** [review]
LGTM, with some inline suggestions. I think this is overall good, however seeing getNextNode etc associated with DPValues makes me twitch, it's a pavlovian response as getNextNode is almost always associated with a gnochange issue in LLVM-IR. I don't think we'll keep this interface forever though, so this should be fine.

File: llvm/lib/IR/DebugProgramInstruction.cpp (hunk 1 above)
Comment (SLTozer) [reply llvm/lib/IR/DebugProgramInstruction.cpp:51]: Perhaps - there might be performance concerns here though: allowing `BasicBlock::iterator` to be passed as an argument here would require `DebugProgramInstruction.h` to `#include "llvm/IR/Instruction.h"`, which _might_ impact performance (including `DebugInfoMetadata.h` here caused a `+2.10%` compile time increase entirely by itself!). I agree that we should probably take a BasicBlock::iterator if possible though; would you consider it acceptable to simply remove the `Instruction *InsertBefore` version (the `DPValue* InsertBefore` is the one that is actually _needed_ for these patches, this overload is optional) and add an iterator-taking version when/if it's confirmed to work, or some other workaround is established?

Comment (jmorse) [reply llvm/lib/IR/DebugProgramInstruction.cpp:51]: You miiiggghht be able to predeclare it as `SymbolTableList<Instruction, ilist_iterator_bits<true>>::iterator;` without needing a body definition for Instruction.

Removing and adding when necessary seems alright -- hopefully we can avoid it being necessary.

Comment (SLTozer) [reply llvm/lib/IR/DebugProgramInstruction.cpp:51]: I tried that first, but Instruction.h defines a template specialization for Instruction, so unfortunately the predeclaration doesn't work.

**Pull request merged**